### PR TITLE
FIR IDE: AddWhenRemainingBranchFix

### PR DESCRIPTION
**Pull request by tgeng**
The first 5 commits are identical with those in \#4427

The fix reuses logic that is already available from
FirWhenExhaustivenessTransformer to collect missing when branches. The
current logic unfortunately uses hackyAllowRunningOnEdt to shorten the
generated code.

In addition, commit#4 makes FIR accept `when(nothing){}` by introducing
`TRIVIALLY_EXHAUSTIVE` status.

### CODE REVIEW

File: compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/FirWhenExhaustivenessTransformer.kt [hunk 1]
```diff
@@ -33,6 +33,58 @@ class FirWhenExhaustivenessTransformer(private val bodyResolveComponents: BodyRe
...
+            subjectType: ConeKotlinType
+        ): List<WhenExhaustivenessChecker> {
+            return buildList {
+                exhaustivenessCheckers.filterTo<WhenExhaustivenessChecker, MutableCollection<in WhenExhaustivenessChecker>>(this) {
+                    it.isApplicable(
+                        subjectType,
+                        session
+                    )
+                }
+                if (isNotEmpty() && subjectType.isMarkedNullable) {
```
[the remark this answers is not in the document]
Comment (tgeng) [reply compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/FirWhenExhaustivenessTransformer.kt:66]: I am guessing if the subject is not a boolean, enum, or sealed class, then the only branch should be an `else` branch, regardless of whether the subject is nullable or not. If we don't have the emptiness check, then the diagnostic/inspection would suggest to create a `null` branch when the subject is an arbitrary class.

File: compiler/testData/diagnostics/tests/controlStructures/whenWithNothingTypedSubject.fir.kt [hunk 2]
```diff
@@ -0,0 +1,8 @@
+// !DIAGNOSTICS: -UNREACHABLE_CODE
+
+typealias MyNothing = Nothing
+
+fun foo(n: Nothing, n2: MyNothing) {
+    val a: Unit = when(n) {}
+//    val b: Unit = when(n2) {}
```
Comment (punzki) [inline compiler/testData/diagnostics/tests/controlStructures/whenWithNothingTypedSubject.fir.kt:7]: The FIR and FE 1.0 files would be identical except that here you have this line commented out. I don't know if you meant to just delete the line completely. Don't think you need it.

File: idea/idea-frontend-api/src/org/jetbrains/kotlin/idea/frontend/api/components/KtReferenceShortener.kt [hunk 3]
```diff
@@ -35,22 +36,40 @@ abstract class KtReferenceShortener : KtAnalysisSessionComponent() {
 }
 
 interface KtReferenceShortenerMixIn : KtAnalysisSessionMixIn {
-    fun collectPossibleReferenceShortenings(
-        file: KtFile,
-        selection: TextRange = file.textRange,
-        classShortenOption: (KtClassOrObjectSymbol) -> ShortenOption = {
+    companion object {
+        val defaultClassShortenOption: (KtClassOrObjectSymbol) -> ShortenOption = {
```
Comment (punzki) [inline idea/idea-frontend-api/src/org/jetbrains/kotlin/idea/frontend/api/components/KtReferenceShortener.kt:40]: These 2 properties can be private (unless you were planning on using it somewhere else)

File: idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddWhenRemainingBranchFixFactories.kt [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+                        })
+
+                }
+            }
+            shortenCommand.invokeShortening()
+        }
+    }
+
+    @OptIn(ExperimentalStdlibApi::class)
+    val noElseInWhen = diagnosticFixFactory<KtFirDiagnostic.NoElseInWhen> { diagnostic ->
```
Comment (punzki) [inline idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddWhenRemainingBranchFixFactories.kt:58]: Please rebase as I changed the API of this (diagnostic class needs to be passed instead of type argument)

File: compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/FirWhenExhaustivenessTransformer.kt [hunk 5]
```diff
@@ -33,6 +33,62 @@ class FirWhenExhaustivenessTransformer(private val bodyResolveComponents: BodyRe
...
+                    collectMissingCases(checkers, whenExpression, type, session)
+                }
+            }
+        }
+
+        private fun getSubjectType(session: FirSession, whenExpression: FirWhenExpression): ConeKotlinType? {
+            val subjectType = whenExpression.subjectVariable?.returnTypeRef?.coneType
+                ?: whenExpression.subject?.typeRef?.coneType
+                ?: run {
+                    whenExpression.replaceExhaustivenessStatus(ExhaustivenessStatus.NotExhaustive.NO_ELSE_BRANCH)
```
Comment (punzki) [inline compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/FirWhenExhaustivenessTransformer.kt:52]: Why are we doing this here? We're just getting the subject type, not doing the actual check. This should be kept in `processExhaustivenessCheck` in line 111 before returning.

File: compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/FirWhenExhaustivenessTransformer.kt [hunk 6]
```diff
@@ -33,6 +33,62 @@ class FirWhenExhaustivenessTransformer(private val bodyResolveComponents: BodyRe
...
+                    return null
+                }
+
+            return subjectType.fullyExpandedType(session).lowerBoundIfFlexible()
+        }
+
+        private fun ConeKotlinType.unwrap(): Collection<ConeKotlinType> {
```
Comment (punzki) [inline compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/FirWhenExhaustivenessTransformer.kt:59]: Maybe call this `unwrapIntersectionTypes()`?

File: idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddWhenRemainingBranchFixFactories.kt [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+import org.jetbrains.kotlin.idea.frontend.api.fir.diagnostics.KtFirDiagnostic
+import org.jetbrains.kotlin.idea.frontend.api.symbols.KtClassKind
+import org.jetbrains.kotlin.idea.frontend.api.tokens.HackToForceAllowRunningAnalyzeOnEDT
+import org.jetbrains.kotlin.idea.frontend.api.tokens.hackyAllowRunningOnEdt
+import org.jetbrains.kotlin.idea.util.generateWhenBranches
+import org.jetbrains.kotlin.name.ClassId
+import org.jetbrains.kotlin.psi.KtWhenExpression
+
+object AddWhenRemainingBranchFixFactories {
+    class Input(val whenMissingCases: List<WhenMissingCase>, val enumsToStarImport: Set<ClassId>) : HLApplicatorInput
```
Comment (punzki) [inline idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddWhenRemainingBranchFixFactories.kt:26]: There's only ever one enum to star import so `enumsToStarImport` can be `ClassId?`

File: compiler/testData/diagnostics/tests/controlStructures/whenWithNothingTypedSubject.fir.kt (hunk 2 above)
Comment (tgeng) [reply compiler/testData/diagnostics/tests/controlStructures/whenWithNothingTypedSubject.fir.kt:7]: Oops! Let me remove this file.

File: idea/idea-frontend-api/src/org/jetbrains/kotlin/idea/frontend/api/components/KtReferenceShortener.kt (hunk 3 above)
Comment (tgeng) [reply idea/idea-frontend-api/src/org/jetbrains/kotlin/idea/frontend/api/components/KtReferenceShortener.kt:40]: Thanks! Making them private sounds good.

File: compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/FirWhenExhaustivenessTransformer.kt (hunk 5 above)
Comment (tgeng) [reply compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/FirWhenExhaustivenessTransformer.kt:52]: Thanks! Good point!

File: idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddWhenRemainingBranchFixFactories.kt (hunk 7 above)
Comment (tgeng) [reply idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddWhenRemainingBranchFixFactories.kt:26]: Sounds good!

File: .idea/vcs.xml [hunk 8]
```diff
@@ -8,18 +8,6 @@
       <inspection_tool class="SubjectLimit" enabled="true" level="WARNING" enabled_by_default="true" />
     </profile>
   </component>
-  <component name="GithubSharedProjectSettings">
```
Comment (jsjeon) [inline .idea/vcs.xml:11]: Are these related changes/removals?

Comment (tgeng) [reply .idea/vcs.xml]: Oops! Somehow my IDE insists changing this. Let me revert it.

File: idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddWhenRemainingBranchFixFactories.kt [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+                }
+            }
+            shortenCommand.invokeShortening()
+        }
+    }
+
+    @OptIn(ExperimentalStdlibApi::class)
+    val noElseInWhen = diagnosticFixFactory(KtFirDiagnostic.NoElseInWhen::class) { diagnostic ->
+        val whenExpression = diagnostic.psi
+        if (whenExpression.closeBrace == null) return@diagnosticFixFactory emptyList()
```
Comment (darthorimar) [inline idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddWhenRemainingBranchFixFactories.kt:60]: Maybe we just can generate it? what if the code is following `when(expression)` -- we can generate braces in such case

File: idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddWhenRemainingBranchFixFactories.kt [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+            else KotlinBundle.lazyMessage("fix.add.remaining.branches")
+        )
+        applyTo { whenExpression, input ->
+            generateWhenBranches(whenExpression, input.whenMissingCases)
+            val shortenCommand = hackyAllowRunningOnEdt {
+                analyse(whenExpression) {
+                    collectPossibleReferenceShorteningsInElement(
+                        whenExpression,
+                        callableShortenOption = {
+                            if (useStarImport && it.callableIdIfNonLocal?.classId == input.enumToStarImport) {
```
Comment (punzki) [inline idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddWhenRemainingBranchFixFactories.kt:44]: You should also check that `input.enumToStarImport != null`

Or just check in the beginning of `applyTo`: `if (useStarImport && input.enumToStarImport == null) return@applyTo`

Comment (tgeng) [reply idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddWhenRemainingBranchFixFactories.kt:44]: Thanks! Let me add an assertion for that.

File: idea/idea-frontend-independent/src/org/jetbrains/kotlin/idea/util/psiUtils.kt [hunk 11]
```diff
@@ -52,3 +53,8 @@ fun PsiClass.classIdIfNonLocal(): ClassId? {
     return ClassId(packageFqName, FqName(classesNames.joinToString(separator = ".")), false)
 }
 
+fun FqName.quoteSegmentsIfNeeded(): String {
```
[the remark this answers is not in the document]
Comment (tgeng) [reply idea/idea-frontend-independent/src/org/jetbrains/kotlin/idea/util/psiUtils.kt]: Thanks! Sounds good!

File: idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/MainKtQuickFixRegistrar.kt [hunk 12]
```diff
@@ -132,6 +132,8 @@ class MainKtQuickFixRegistrar : KtQuickFixRegistrar() {
         registerApplicator(WrapWithSafeLetCallFixFactories.forUnsafeInfixCall)
         registerApplicator(WrapWithSafeLetCallFixFactories.forUnsafeOperatorCall)
         registerApplicator(WrapWithSafeLetCallFixFactories.forArgumentTypeMismatch)
+
+        registerApplicator(AddWhenRemainingBranchFixFactories.noElseInWhen)
```
[the remark this answers is not in the document]
Comment (tgeng) [reply idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/MainKtQuickFixRegistrar.kt:137]: OK, let me create a separate group for these.

File: idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddWhenRemainingBranchFixFactories.kt (hunk 9 above)
Comment (tgeng) [reply idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddWhenRemainingBranchFixFactories.kt]: Yep, sounds good! Also added tests covering these cases.

File: idea/testData/quickfix/when/addRemainingBranchesMissingLeftBracket.fir.kt [hunk 13]
```diff
@@ -0,0 +1,10 @@
+// "Add remaining branches" "true"
+// WITH_RUNTIME
+
+sealed class A
+class B : A()
+
+fun test(a: A) {
+  val i = w<caret>hen (a)
+
+/* IGNORE_FIR */
```
Comment (punzki) [inline idea/testData/quickfix/when/addRemainingBranchesMissingLeftBracket.fir.kt:10]: Why do these 3 new tests have `IGNORE_FIR` in the `.fir.kt` file?

Comment (tgeng) [reply idea/testData/quickfix/when/addRemainingBranchesMissingLeftBracket.fir.kt:10]: Oops!

**Review by punzki — APPROVED** [review]

File: compiler/fir/tree/src/org/jetbrains/kotlin/fir/expressions/ExhaustivenessStatus.kt [hunk 14]
```diff
@@ -8,7 +8,19 @@ package org.jetbrains.kotlin.fir.expressions
...
+     * This value is used if the subject has type other than `Nothing`, in which case it's literally exhaustive only if type's possible
+     * cases are properly covered.
+     */
+    object ProperlyExhaustive : ExhaustivenessStatus()
+
+    /**
+     *  This value is used if the subject has type `Nothing`, in which case even an empty `when` is considered exhaustive. Also, in this
+     *  case, a synthetic else branch is created.
+     */
+    object TriviallyExhaustive : ExhaustivenessStatus()
```
Comment (darthorimar) [inline compiler/fir/tree/src/org/jetbrains/kotlin/fir/expressions/ExhaustivenessStatus.kt:22]: May be ExhaustiveAsNothing is a better name?

File: idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddWhenRemainingBranchFixFactories.kt [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+        }
+    }
+
+    @OptIn(ExperimentalStdlibApi::class)
+    val noElseInWhen = diagnosticFixFactory(KtFirDiagnostic.NoElseInWhen::class) { diagnostic ->
+        val whenExpression = diagnostic.psi
+        val subjectExpression = whenExpression.subjectExpression ?: return@diagnosticFixFactory emptyList()
+
+        buildList {
+            val missingCases = whenExpression.getMissingCases().takeIf {
```
Comment (darthorimar) [inline idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddWhenRemainingBranchFixFactories.kt:64]: Why don't we get missing cases from `NoElseInWhen` diagnosic? it contains `NoElseInWhen.missingWhenCases` property

File: compiler/fir/tree/src/org/jetbrains/kotlin/fir/expressions/ExhaustivenessStatus.kt (hunk 14 above)
Comment (tgeng) [reply compiler/fir/tree/src/org/jetbrains/kotlin/fir/expressions/ExhaustivenessStatus.kt:22]: Sure, sounds good!

File: idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddWhenRemainingBranchFixFactories.kt (hunk 15 above)
Comment (tgeng) [reply idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddWhenRemainingBranchFixFactories.kt:64]: Oh... I missed that...
So do you think if it's worth to keeping the high level API to create missing cases from a when expression? Maybe it can be useful for a future intention to generate them for non-sealed when expressions (since there probably won't be any diagnostics for it)?

Comment (darthorimar) [reply idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddWhenRemainingBranchFixFactories.kt]: Sure, let's keep it. But we should check that it will work for when used as non-expression. Mind adding tests that check that?

Comment (tgeng) [reply idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddWhenRemainingBranchFixFactories.kt]: OK, added HLAddWhenRemainingBranchesIntention in the last commit and enabled the corresponding intention test.

**Review by darthorimar — APPROVED** [review]

**Comment by tgeng** [comment]
Hi @darthorimar,  do you prefer me to create a separate PR for intellij-kotlin instead?

**Comment by darthorimar** [comment]
The PR is merged manually into kotlin-ide repository, thanks!
